### ISSUE TITLE
Should ECR provide support for escaping rendered output?

### ISSUE THREAD

**Issue by robacarp** [crystal-lang/crystal#15078]
The docs for ECR state: 

> Embedded Crystal (ECR) is a template language for embedding Crystal code into other text, that includes but is not limited to HTML.

A naïve user (me) might assume that ECR is smart enough to escape syntax so that it doesn't end up causing an XSS vector on a HTML based project. My workaround for this has been laborious, and it includes manually sanitizing anything being rendered with a call to HTML.escape. The workflow is error prone because it requires me to _remember_ to be a good steward of user input. In todays world, web frameworks have evolved to do this for me because it's been shown over and over again that I will fail to remember to do this 100% of the time.

This has led me to consider a bunch of web specific templating languages. Unfortunately there's a lot of bit-rot and abandoned projects in the shardverse, and I'm skiddish to attach my project to another which would require a rewrite when the author no longer has time for a passion project. So I keep coming back to ECR because it's built-in to crystal. 

**Would it be possible to upgrade ECR so that it can take a sanitation helper?** 

I'm humbly attempting to survey the ECR Processor for an easy victory. It seems like maybe [wrapping this `buffer_name`](https://github.com/crystal-lang/crystal/blob/master/src/ecr/processor.cr#L44) in a call to a theoretical `user_escape()` method would suffice -- or maybe it would be elsewhere, I can't tell.

I'm inspired by the syntax for `Log` -- and I love it. `Log = ::Log.for(self)` is clean and elegant.

I'm imagining something like this, which would require a broader rework of ECR, but could be really powerful:

```crystal
HtmlSafeECR = ECR.with_escape do |string|
  HTML.escape string
end

HtmlSafeECR.render template.ecr
```

`ECR.with_escape` would capture the block and apply it every time a `<%= %>` sequence is evaluated, providing a safe _default_ render macro. 

Instructing the escape block that the string is _already_ escaped and shouldn't be handled gets tricky. Perl had `taint` and I remember using that to accomplish this on some protoframework. Ruby has both `taint` and `trust` -- though Rails implements [something completely different](https://api.rubyonrails.org/classes/ActiveSupport/SafeBuffer.html)https://api.rubyonrails.org/classes/ActiveSupport/SafeBuffer.html) for it's usage of ERB.

What do you think?

**Comment by Blacksmoke16** [comment]
~~Duplicate of https://github.com/crystal-lang/crystal/issues/13753?~~

NVM

**Comment by robacarp** [comment]
Another approach which might be more helpful is to make ECR a generic, and the T is the buffer. So the current implementation is:

```crystal
module ECR
  macro render(filename)
    ::String.build do |%io|
      ::ECR.embed({{filename}}, %io)
    end
  end
end
```

But it could also work like this:

```crystal
module ECR(BufferClass)
  macro render(filename)
    BufferClass.build do |%io|
      ::ECR.embed({{filename}}, %io)
    end
  end
end
```

BufferClass must then have a build method, which yields an object (a buffer ingest object). The buffer ingest object must respond to `<<`.

That would allow more in depth inspection of safe/unsafe strings through some out-of-band syntax. The buffer ingest object could be made responsible for deciding what needs to be escaped and what doesn't.

**Comment by robacarp** [comment]
Also somewhat related is that ERB (or Rail's extension of ERB) has a `<%==` operator which _doesn't_ run the html sanitation step. It's subtle, maybe even too subtle. Enough that there are [suggestions to avoid it](https://andycroll.com/ruby/beware-of-raw-erb/).

**Comment by nobodywasishere** [comment]
Even if it's a breaking change, I'd be want to make `<%=` escape html by default and add `<%==` for not escaping, with maybe adding an ameba rule to discourage using the latter.

**Comment by jgaskins** [comment]
> make `<%=` escape html by default and add `<%==` for not escaping

This is precisely what I did with [`Armature::Template`](https://github.com/jgaskins/armature/blob/f1d3dcce1a7214d5fd2e34e2b999483e8d756440/src/template.cr#L183-L190), which was basically a copy/paste of `ECR` with a few extensions made to it, like this one for sanitization. `<%=` passes a `Armature::Template::HTML::SanitizableValue` that wraps the expression, so …

```ecr
<%= user.name %>
```

… is equivalent to …

```ecr
<%== Armature::Template::HTML::SanitizableValue.new(user.name) %>
```

With that said, though, I think it's important to keep `ECR` as a generic embedded Crystal and would maybe prefer some sort of processor plugin similar to what @robacarp suggested to modify the output. Not strictly sanitize/escape, but allow for any modifications as the values pass through the pipeline of processors.

More generally speaking, I don't think `ECR` should assume it's writing HTML output. It's likely the most common use case, it might even be the majority, but I don't think it should be assumed.

That's the reason I used ECR to create `Armature::Template` instead of trying to send a PR for HTML escaping in ECR. `Armature::Template` is part of a web framework so it can focus exclusively on the HTML use case, but ECR is explicitly unconstrained in how it can be used.

**Comment by ysbaddaden** [comment]
I agree with @jgaskins. HTML is only one of the possible formats (MD, YAML, ...)
